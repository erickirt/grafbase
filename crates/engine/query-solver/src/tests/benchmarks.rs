use crate::assert_solving_snapshots;

const SCHEMA: &str = r#"
enum join__Graph {
  SUB0 @join__graph(name: "sub0", url: "http://localhost:1000")
  SUB1 @join__graph(name: "sub1", url: "http://localhost:1001")
  SUB2 @join__graph(name: "sub2", url: "http://localhost:1002")
}

type Query @join__type(graph: SUB0) @join__type(graph: SUB1) @join__type(graph: SUB2) {
    node: Node
}
type Node @join__type(graph: SUB0, key: "id0") @join__type(graph: SUB1, key: "id0") @join__type(graph: SUB2, key: "id0") {
    id0: ID!
    n0: Node @join__field(graph: SUB1) @join__field(graph: SUB2)
    f0: String @join__field(graph: SUB1) @join__field(graph: SUB2)
    n1: Node @join__field(graph: SUB0) @join__field(graph: SUB2)
    f1: String @join__field(graph: SUB0) @join__field(graph: SUB2)
    n2: Node @join__field(graph: SUB0) @join__field(graph: SUB1)
    f2: String @join__field(graph: SUB0) @join__field(graph: SUB1)
}
"#;

#[tokio::test]
async fn alternating_subgraphs_1() {
    assert_solving_snapshots!(
        "alternating_subgraphs_1",
        SCHEMA,
        r#"
        query {
          node {
            n0 {
              n0 {
                f0
                f1
                f2
              }
              n2 {
                f0
                f1
                f2
              }
            }
          }
        }
        "#
    );
}
