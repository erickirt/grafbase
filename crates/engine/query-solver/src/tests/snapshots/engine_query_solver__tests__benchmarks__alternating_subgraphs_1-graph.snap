---
source: crates/engine/query-solver/src/tests/benchmarks.rs
expression: "digraph {\n    0 [ label = \"root\" ]\n    1 [ label = \"Query.node\" ]\n    2 [ label = \"Node.n0\" ]\n    3 [ label = \"Node.n0\" ]\n    4 [ label = \"Node.n2\" ]\n    5 [ label = \"Node.f1\" ]\n    6 [ label = \"Node.f2\" ]\n    7 [ label = \"Node.f0\" ]\n    8 [ label = \"Node.f1\" ]\n    9 [ label = \"Root#sub0\", shape=parallelogram, color=dodgerblue ]\n    10 [ label = \"node#sub0\", shape=box, color=dodgerblue ]\n    11 [ label = \"Root#sub1\", shape=parallelogram, color=dodgerblue ]\n    12 [ label = \"node#sub1\", shape=box, color=dodgerblue ]\n    13 [ label = \"Root#sub2\", shape=parallelogram, color=dodgerblue ]\n    14 [ label = \"node#sub2\", shape=box, color=dodgerblue ]\n    15 [ label = \"n0#sub2\", shape=box, color=dodgerblue ]\n    16 [ label = \"FedEntity#sub1\", shape=parallelogram, color=dodgerblue ]\n    17 [ label = \"n0#sub1\", shape=box, color=dodgerblue ]\n    18 [ label = \"n0#sub1\", shape=box, color=dodgerblue ]\n    19 [ label = \"FedEntity#sub2\", shape=parallelogram, color=dodgerblue ]\n    20 [ label = \"n0#sub2\", shape=box, color=dodgerblue ]\n    21 [ label = \"f1#sub2\", shape=box, color=dodgerblue ]\n    22 [ label = \"FedEntity#sub0\", shape=parallelogram, color=dodgerblue ]\n    23 [ label = \"f1#sub0\", shape=box, color=dodgerblue ]\n    24 [ label = \"f2#sub0\", shape=box, color=dodgerblue ]\n    25 [ label = \"FedEntity#sub1\", shape=parallelogram, color=dodgerblue ]\n    26 [ label = \"f2#sub1\", shape=box, color=dodgerblue ]\n    27 [ label = \"FedEntity#sub2\", shape=parallelogram, color=dodgerblue ]\n    28 [ label = \"f1#sub2\", shape=box, color=dodgerblue ]\n    29 [ label = \"f2#sub1\", shape=box, color=dodgerblue ]\n    30 [ label = \"n2#sub1\", shape=box, color=dodgerblue ]\n    31 [ label = \"FedEntity#sub0\", shape=parallelogram, color=dodgerblue ]\n    32 [ label = \"n2#sub0\", shape=box, color=dodgerblue ]\n    33 [ label = \"FedEntity#sub1\", shape=parallelogram, color=dodgerblue ]\n    34 [ label = \"f0#sub1\", shape=box, color=dodgerblue ]\n    35 [ label = \"FedEntity#sub2\", shape=parallelogram, color=dodgerblue ]\n    36 [ label = \"f0#sub2\", shape=box, color=dodgerblue ]\n    37 [ label = \"f1#sub0\", shape=box, color=dodgerblue ]\n    38 [ label = \"f1#sub2\", shape=box, color=dodgerblue ]\n    39 [ label = \"f0#sub1\", shape=box, color=dodgerblue ]\n    40 [ label = \"FedEntity#sub0\", shape=parallelogram, color=dodgerblue ]\n    41 [ label = \"f1#sub0\", shape=box, color=dodgerblue ]\n    42 [ label = \"n0#sub2\", shape=box, color=dodgerblue ]\n    43 [ label = \"FedEntity#sub1\", shape=parallelogram, color=dodgerblue ]\n    44 [ label = \"n0#sub1\", shape=box, color=dodgerblue ]\n    45 [ label = \"f2#sub1\", shape=box, color=dodgerblue ]\n    46 [ label = \"f1#sub2\", shape=box, color=dodgerblue ]\n    47 [ label = \"n2#sub1\", shape=box, color=dodgerblue ]\n    48 [ label = \"f0#sub1\", shape=box, color=dodgerblue ]\n    49 [ label = \"n0#sub1\", shape=box, color=dodgerblue ]\n    50 [ label = \"FedEntity#sub2\", shape=parallelogram, color=dodgerblue ]\n    51 [ label = \"n0#sub2\", shape=box, color=dodgerblue ]\n    52 [ label = \"n0#sub2\", shape=box, color=dodgerblue ]\n    53 [ label = \"f1#sub2\", shape=box, color=dodgerblue ]\n    54 [ label = \"n0#sub1\", shape=box, color=dodgerblue ]\n    55 [ label = \"f2#sub1\", shape=box, color=dodgerblue ]\n    56 [ label = \"n2#sub1\", shape=box, color=dodgerblue ]\n    57 [ label = \"f0#sub1\", shape=box, color=dodgerblue ]\n    58 [ label = \"*Node.id0\" ]\n    59 [ label = \"id0#sub0\", shape=box, color=dodgerblue ]\n    60 [ label = \"id0#sub1\", shape=box, color=dodgerblue ]\n    61 [ label = \"id0#sub2\", shape=box, color=dodgerblue ]\n    62 [ label = \"*Node.id0\" ]\n    63 [ label = \"id0#sub2\", shape=box, color=dodgerblue ]\n    64 [ label = \"id0#sub1\", shape=box, color=dodgerblue ]\n    65 [ label = \"id0#sub1\", shape=box, color=dodgerblue ]\n    66 [ label = \"id0#sub2\", shape=box, color=dodgerblue ]\n    67 [ label = \"*Node.id0\" ]\n    68 [ label = \"id0#sub1\", shape=box, color=dodgerblue ]\n    69 [ label = \"id0#sub0\", shape=box, color=dodgerblue ]\n    70 [ label = \"id0#sub1\", shape=box, color=dodgerblue ]\n    71 [ label = \"id0#sub1\", shape=box, color=dodgerblue ]\n    72 [ label = \"*Node.id0\" ]\n    73 [ label = \"id0#sub1\", shape=box, color=dodgerblue ]\n    74 [ label = \"id0#sub2\", shape=box, color=dodgerblue ]\n    75 [ label = \"id0#sub2\", shape=box, color=dodgerblue ]\n    76 [ label = \"id0#sub1\", shape=box, color=dodgerblue ]\n    77 [ label = \"id0#sub2\", shape=box, color=dodgerblue ]\n    78 [ label = \"id0#sub1\", shape=box, color=dodgerblue ]\n    0 -> 1 [  ]\n    1 -> 2 [  ]\n    2 -> 3 [  ]\n    2 -> 4 [  ]\n    3 -> 5 [  ]\n    3 -> 6 [  ]\n    4 -> 7 [  ]\n    4 -> 8 [  ]\n    0 -> 9 [ color=royalblue,fontcolor=royalblue ]\n    0 -> 9 [ style=dashed,arrowhead=none ]\n    9 -> 10 [ color=royalblue,fontcolor=royalblue ]\n    10 -> 1 [ color=violet,arrowhead=none ]\n    0 -> 11 [ color=royalblue,fontcolor=royalblue ]\n    0 -> 11 [ style=dashed,arrowhead=none ]\n    11 -> 12 [ color=royalblue,fontcolor=royalblue ]\n    12 -> 1 [ color=violet,arrowhead=none ]\n    0 -> 13 [ color=royalblue,fontcolor=royalblue ]\n    0 -> 13 [ style=dashed,arrowhead=none ]\n    13 -> 14 [ color=royalblue,fontcolor=royalblue ]\n    14 -> 1 [ color=violet,arrowhead=none ]\n    14 -> 15 [ color=royalblue,fontcolor=royalblue ]\n    15 -> 2 [ color=violet,arrowhead=none ]\n    14 -> 16 [ color=royalblue,fontcolor=royalblue ]\n    1 -> 16 [ style=dashed,arrowhead=none ]\n    16 -> 17 [ color=royalblue,fontcolor=royalblue ]\n    17 -> 2 [ color=violet,arrowhead=none ]\n    17 -> 18 [ color=royalblue,fontcolor=royalblue ]\n    18 -> 3 [ color=violet,arrowhead=none ]\n    17 -> 19 [ color=royalblue,fontcolor=royalblue ]\n    2 -> 19 [ style=dashed,arrowhead=none ]\n    19 -> 20 [ color=royalblue,fontcolor=royalblue ]\n    20 -> 3 [ color=violet,arrowhead=none ]\n    20 -> 21 [ color=royalblue,fontcolor=royalblue ]\n    21 -> 5 [ color=violet,arrowhead=none ]\n    20 -> 22 [ color=royalblue,fontcolor=royalblue ]\n    3 -> 22 [ style=dashed,arrowhead=none ]\n    22 -> 23 [ color=royalblue,fontcolor=royalblue ]\n    23 -> 5 [ color=violet,arrowhead=none ]\n    22 -> 24 [ color=royalblue,fontcolor=royalblue ]\n    24 -> 6 [ color=violet,arrowhead=none ]\n    20 -> 25 [ color=royalblue,fontcolor=royalblue ]\n    3 -> 25 [ style=dashed,arrowhead=none ]\n    25 -> 26 [ color=royalblue,fontcolor=royalblue ]\n    26 -> 6 [ color=violet,arrowhead=none ]\n    18 -> 22 [ color=royalblue,fontcolor=royalblue ]\n    18 -> 27 [ color=royalblue,fontcolor=royalblue ]\n    3 -> 27 [ style=dashed,arrowhead=none ]\n    27 -> 28 [ color=royalblue,fontcolor=royalblue ]\n    28 -> 5 [ color=violet,arrowhead=none ]\n    18 -> 29 [ color=royalblue,fontcolor=royalblue ]\n    29 -> 6 [ color=violet,arrowhead=none ]\n    17 -> 30 [ color=royalblue,fontcolor=royalblue ]\n    30 -> 4 [ color=violet,arrowhead=none ]\n    17 -> 31 [ color=royalblue,fontcolor=royalblue ]\n    2 -> 31 [ style=dashed,arrowhead=none ]\n    31 -> 32 [ color=royalblue,fontcolor=royalblue ]\n    32 -> 4 [ color=violet,arrowhead=none ]\n    32 -> 33 [ color=royalblue,fontcolor=royalblue ]\n    4 -> 33 [ style=dashed,arrowhead=none ]\n    33 -> 34 [ color=royalblue,fontcolor=royalblue ]\n    34 -> 7 [ color=violet,arrowhead=none ]\n    32 -> 35 [ color=royalblue,fontcolor=royalblue ]\n    4 -> 35 [ style=dashed,arrowhead=none ]\n    35 -> 36 [ color=royalblue,fontcolor=royalblue ]\n    36 -> 7 [ color=violet,arrowhead=none ]\n    32 -> 37 [ color=royalblue,fontcolor=royalblue ]\n    37 -> 8 [ color=violet,arrowhead=none ]\n    35 -> 38 [ color=royalblue,fontcolor=royalblue ]\n    38 -> 8 [ color=violet,arrowhead=none ]\n    30 -> 39 [ color=royalblue,fontcolor=royalblue ]\n    39 -> 7 [ color=violet,arrowhead=none ]\n    30 -> 35 [ color=royalblue,fontcolor=royalblue ]\n    30 -> 40 [ color=royalblue,fontcolor=royalblue ]\n    4 -> 40 [ style=dashed,arrowhead=none ]\n    40 -> 41 [ color=royalblue,fontcolor=royalblue ]\n    41 -> 8 [ color=violet,arrowhead=none ]\n    15 -> 42 [ color=royalblue,fontcolor=royalblue ]\n    42 -> 3 [ color=violet,arrowhead=none ]\n    15 -> 43 [ color=royalblue,fontcolor=royalblue ]\n    2 -> 43 [ style=dashed,arrowhead=none ]\n    43 -> 44 [ color=royalblue,fontcolor=royalblue ]\n    44 -> 3 [ color=violet,arrowhead=none ]\n    44 -> 22 [ color=royalblue,fontcolor=royalblue ]\n    44 -> 27 [ color=royalblue,fontcolor=royalblue ]\n    44 -> 45 [ color=royalblue,fontcolor=royalblue ]\n    45 -> 6 [ color=violet,arrowhead=none ]\n    42 -> 46 [ color=royalblue,fontcolor=royalblue ]\n    46 -> 5 [ color=violet,arrowhead=none ]\n    42 -> 22 [ color=royalblue,fontcolor=royalblue ]\n    42 -> 25 [ color=royalblue,fontcolor=royalblue ]\n    15 -> 31 [ color=royalblue,fontcolor=royalblue ]\n    43 -> 47 [ color=royalblue,fontcolor=royalblue ]\n    47 -> 4 [ color=violet,arrowhead=none ]\n    47 -> 48 [ color=royalblue,fontcolor=royalblue ]\n    48 -> 7 [ color=violet,arrowhead=none ]\n    47 -> 35 [ color=royalblue,fontcolor=royalblue ]\n    47 -> 40 [ color=royalblue,fontcolor=royalblue ]\n    12 -> 49 [ color=royalblue,fontcolor=royalblue ]\n    49 -> 2 [ color=violet,arrowhead=none ]\n    12 -> 50 [ color=royalblue,fontcolor=royalblue ]\n    1 -> 50 [ style=dashed,arrowhead=none ]\n    50 -> 51 [ color=royalblue,fontcolor=royalblue ]\n    51 -> 2 [ color=violet,arrowhead=none ]\n    51 -> 52 [ color=royalblue,fontcolor=royalblue ]\n    52 -> 3 [ color=violet,arrowhead=none ]\n    51 -> 43 [ color=royalblue,fontcolor=royalblue ]\n    52 -> 53 [ color=royalblue,fontcolor=royalblue ]\n    53 -> 5 [ color=violet,arrowhead=none ]\n    52 -> 22 [ color=royalblue,fontcolor=royalblue ]\n    52 -> 25 [ color=royalblue,fontcolor=royalblue ]\n    51 -> 31 [ color=royalblue,fontcolor=royalblue ]\n    49 -> 54 [ color=royalblue,fontcolor=royalblue ]\n    54 -> 3 [ color=violet,arrowhead=none ]\n    49 -> 19 [ color=royalblue,fontcolor=royalblue ]\n    54 -> 22 [ color=royalblue,fontcolor=royalblue ]\n    54 -> 27 [ color=royalblue,fontcolor=royalblue ]\n    54 -> 55 [ color=royalblue,fontcolor=royalblue ]\n    55 -> 6 [ color=violet,arrowhead=none ]\n    49 -> 56 [ color=royalblue,fontcolor=royalblue ]\n    56 -> 4 [ color=violet,arrowhead=none ]\n    49 -> 31 [ color=royalblue,fontcolor=royalblue ]\n    56 -> 57 [ color=royalblue,fontcolor=royalblue ]\n    57 -> 7 [ color=violet,arrowhead=none ]\n    56 -> 35 [ color=royalblue,fontcolor=royalblue ]\n    56 -> 40 [ color=royalblue,fontcolor=royalblue ]\n    10 -> 16 [ color=royalblue,fontcolor=royalblue ]\n    10 -> 50 [ color=royalblue,fontcolor=royalblue ]\n    1 -> 58 [  ]\n    50 -> 58 [ color=orangered,arrowhead=inv ]\n    10 -> 59 [ color=royalblue,fontcolor=royalblue ]\n    59 -> 58 [ color=violet,arrowhead=none ]\n    12 -> 60 [ color=royalblue,fontcolor=royalblue ]\n    60 -> 58 [ color=violet,arrowhead=none ]\n    14 -> 61 [ color=royalblue,fontcolor=royalblue ]\n    61 -> 58 [ color=violet,arrowhead=none ]\n    2 -> 62 [  ]\n    43 -> 62 [ color=orangered,arrowhead=inv ]\n    15 -> 63 [ color=royalblue,fontcolor=royalblue ]\n    63 -> 62 [ color=violet,arrowhead=none ]\n    17 -> 64 [ color=royalblue,fontcolor=royalblue ]\n    64 -> 62 [ color=violet,arrowhead=none ]\n    49 -> 65 [ color=royalblue,fontcolor=royalblue ]\n    65 -> 62 [ color=violet,arrowhead=none ]\n    51 -> 66 [ color=royalblue,fontcolor=royalblue ]\n    66 -> 62 [ color=violet,arrowhead=none ]\n    4 -> 67 [  ]\n    40 -> 67 [ color=orangered,arrowhead=inv ]\n    30 -> 68 [ color=royalblue,fontcolor=royalblue ]\n    68 -> 67 [ color=violet,arrowhead=none ]\n    32 -> 69 [ color=royalblue,fontcolor=royalblue ]\n    69 -> 67 [ color=violet,arrowhead=none ]\n    47 -> 70 [ color=royalblue,fontcolor=royalblue ]\n    70 -> 67 [ color=violet,arrowhead=none ]\n    56 -> 71 [ color=royalblue,fontcolor=royalblue ]\n    71 -> 67 [ color=violet,arrowhead=none ]\n    35 -> 67 [ color=orangered,arrowhead=inv ]\n    33 -> 67 [ color=orangered,arrowhead=inv ]\n    31 -> 62 [ color=orangered,arrowhead=inv ]\n    3 -> 72 [  ]\n    27 -> 72 [ color=orangered,arrowhead=inv ]\n    18 -> 73 [ color=royalblue,fontcolor=royalblue ]\n    73 -> 72 [ color=violet,arrowhead=none ]\n    20 -> 74 [ color=royalblue,fontcolor=royalblue ]\n    74 -> 72 [ color=violet,arrowhead=none ]\n    42 -> 75 [ color=royalblue,fontcolor=royalblue ]\n    75 -> 72 [ color=violet,arrowhead=none ]\n    44 -> 76 [ color=royalblue,fontcolor=royalblue ]\n    76 -> 72 [ color=violet,arrowhead=none ]\n    52 -> 77 [ color=royalblue,fontcolor=royalblue ]\n    77 -> 72 [ color=violet,arrowhead=none ]\n    54 -> 78 [ color=royalblue,fontcolor=royalblue ]\n    78 -> 72 [ color=violet,arrowhead=none ]\n    25 -> 72 [ color=orangered,arrowhead=inv ]\n    22 -> 72 [ color=orangered,arrowhead=inv ]\n    19 -> 62 [ color=orangered,arrowhead=inv ]\n    16 -> 58 [ color=orangered,arrowhead=inv ]\n}\n"
---
digraph {
    0 [ root]
    1 [ Query.node]
    2 [ Node.n0]
    3 [ Node.n0]
    4 [ Node.n2]
    5 [ Node.f1]
    6 [ Node.f2]
    7 [ Node.f0]
    8 [ Node.f1]
    9 [ Root#sub0]
    10 [ node#sub0]
    11 [ Root#sub1]
    12 [ node#sub1]
    13 [ Root#sub2]
    14 [ node#sub2]
    15 [ n0#sub2]
    16 [ FedEntity#sub1]
    17 [ n0#sub1]
    18 [ n0#sub1]
    19 [ FedEntity#sub2]
    20 [ n0#sub2]
    21 [ f1#sub2]
    22 [ FedEntity#sub0]
    23 [ f1#sub0]
    24 [ f2#sub0]
    25 [ FedEntity#sub1]
    26 [ f2#sub1]
    27 [ FedEntity#sub2]
    28 [ f1#sub2]
    29 [ f2#sub1]
    30 [ n2#sub1]
    31 [ FedEntity#sub0]
    32 [ n2#sub0]
    33 [ FedEntity#sub1]
    34 [ f0#sub1]
    35 [ FedEntity#sub2]
    36 [ f0#sub2]
    37 [ f1#sub0]
    38 [ f1#sub2]
    39 [ f0#sub1]
    40 [ FedEntity#sub0]
    41 [ f1#sub0]
    42 [ n0#sub2]
    43 [ FedEntity#sub1]
    44 [ n0#sub1]
    45 [ f2#sub1]
    46 [ f1#sub2]
    47 [ n2#sub1]
    48 [ f0#sub1]
    49 [ n0#sub1]
    50 [ FedEntity#sub2]
    51 [ n0#sub2]
    52 [ n0#sub2]
    53 [ f1#sub2]
    54 [ n0#sub1]
    55 [ f2#sub1]
    56 [ n2#sub1]
    57 [ f0#sub1]
    58 [ *Node.id0]
    59 [ id0#sub0]
    60 [ id0#sub1]
    61 [ id0#sub2]
    62 [ *Node.id0]
    63 [ id0#sub2]
    64 [ id0#sub1]
    65 [ id0#sub1]
    66 [ id0#sub2]
    67 [ *Node.id0]
    68 [ id0#sub1]
    69 [ id0#sub0]
    70 [ id0#sub1]
    71 [ id0#sub1]
    72 [ *Node.id0]
    73 [ id0#sub1]
    74 [ id0#sub2]
    75 [ id0#sub2]
    76 [ id0#sub1]
    77 [ id0#sub2]
    78 [ id0#sub1]
    0 -> 1 [ label = "Field" ]
    1 -> 2 [ label = "Field" ]
    2 -> 3 [ label = "Field" ]
    2 -> 4 [ label = "Field" ]
    3 -> 5 [ label = "Field" ]
    3 -> 6 [ label = "Field" ]
    4 -> 7 [ label = "Field" ]
    4 -> 8 [ label = "Field" ]
    0 -> 9 [ label = "CreateChildResolver" ]
    0 -> 9 [ label = "HasChildResolver" ]
    9 -> 10 [ label = "CanProvide" ]
    10 -> 1 [ label = "Provides" ]
    0 -> 11 [ label = "CreateChildResolver" ]
    0 -> 11 [ label = "HasChildResolver" ]
    11 -> 12 [ label = "CanProvide" ]
    12 -> 1 [ label = "Provides" ]
    0 -> 13 [ label = "CreateChildResolver" ]
    0 -> 13 [ label = "HasChildResolver" ]
    13 -> 14 [ label = "CanProvide" ]
    14 -> 1 [ label = "Provides" ]
    14 -> 15 [ label = "CanProvide" ]
    15 -> 2 [ label = "Provides" ]
    14 -> 16 [ label = "CreateChildResolver" ]
    1 -> 16 [ label = "HasChildResolver" ]
    16 -> 17 [ label = "CanProvide" ]
    17 -> 2 [ label = "Provides" ]
    17 -> 18 [ label = "CanProvide" ]
    18 -> 3 [ label = "Provides" ]
    17 -> 19 [ label = "CreateChildResolver" ]
    2 -> 19 [ label = "HasChildResolver" ]
    19 -> 20 [ label = "CanProvide" ]
    20 -> 3 [ label = "Provides" ]
    20 -> 21 [ label = "CanProvide" ]
    21 -> 5 [ label = "Provides" ]
    20 -> 22 [ label = "CreateChildResolver" ]
    3 -> 22 [ label = "HasChildResolver" ]
    22 -> 23 [ label = "CanProvide" ]
    23 -> 5 [ label = "Provides" ]
    22 -> 24 [ label = "CanProvide" ]
    24 -> 6 [ label = "Provides" ]
    20 -> 25 [ label = "CreateChildResolver" ]
    3 -> 25 [ label = "HasChildResolver" ]
    25 -> 26 [ label = "CanProvide" ]
    26 -> 6 [ label = "Provides" ]
    18 -> 22 [ label = "CreateChildResolver" ]
    18 -> 27 [ label = "CreateChildResolver" ]
    3 -> 27 [ label = "HasChildResolver" ]
    27 -> 28 [ label = "CanProvide" ]
    28 -> 5 [ label = "Provides" ]
    18 -> 29 [ label = "CanProvide" ]
    29 -> 6 [ label = "Provides" ]
    17 -> 30 [ label = "CanProvide" ]
    30 -> 4 [ label = "Provides" ]
    17 -> 31 [ label = "CreateChildResolver" ]
    2 -> 31 [ label = "HasChildResolver" ]
    31 -> 32 [ label = "CanProvide" ]
    32 -> 4 [ label = "Provides" ]
    32 -> 33 [ label = "CreateChildResolver" ]
    4 -> 33 [ label = "HasChildResolver" ]
    33 -> 34 [ label = "CanProvide" ]
    34 -> 7 [ label = "Provides" ]
    32 -> 35 [ label = "CreateChildResolver" ]
    4 -> 35 [ label = "HasChildResolver" ]
    35 -> 36 [ label = "CanProvide" ]
    36 -> 7 [ label = "Provides" ]
    32 -> 37 [ label = "CanProvide" ]
    37 -> 8 [ label = "Provides" ]
    35 -> 38 [ label = "CanProvide" ]
    38 -> 8 [ label = "Provides" ]
    30 -> 39 [ label = "CanProvide" ]
    39 -> 7 [ label = "Provides" ]
    30 -> 35 [ label = "CreateChildResolver" ]
    30 -> 40 [ label = "CreateChildResolver" ]
    4 -> 40 [ label = "HasChildResolver" ]
    40 -> 41 [ label = "CanProvide" ]
    41 -> 8 [ label = "Provides" ]
    15 -> 42 [ label = "CanProvide" ]
    42 -> 3 [ label = "Provides" ]
    15 -> 43 [ label = "CreateChildResolver" ]
    2 -> 43 [ label = "HasChildResolver" ]
    43 -> 44 [ label = "CanProvide" ]
    44 -> 3 [ label = "Provides" ]
    44 -> 22 [ label = "CreateChildResolver" ]
    44 -> 27 [ label = "CreateChildResolver" ]
    44 -> 45 [ label = "CanProvide" ]
    45 -> 6 [ label = "Provides" ]
    42 -> 46 [ label = "CanProvide" ]
    46 -> 5 [ label = "Provides" ]
    42 -> 22 [ label = "CreateChildResolver" ]
    42 -> 25 [ label = "CreateChildResolver" ]
    15 -> 31 [ label = "CreateChildResolver" ]
    43 -> 47 [ label = "CanProvide" ]
    47 -> 4 [ label = "Provides" ]
    47 -> 48 [ label = "CanProvide" ]
    48 -> 7 [ label = "Provides" ]
    47 -> 35 [ label = "CreateChildResolver" ]
    47 -> 40 [ label = "CreateChildResolver" ]
    12 -> 49 [ label = "CanProvide" ]
    49 -> 2 [ label = "Provides" ]
    12 -> 50 [ label = "CreateChildResolver" ]
    1 -> 50 [ label = "HasChildResolver" ]
    50 -> 51 [ label = "CanProvide" ]
    51 -> 2 [ label = "Provides" ]
    51 -> 52 [ label = "CanProvide" ]
    52 -> 3 [ label = "Provides" ]
    51 -> 43 [ label = "CreateChildResolver" ]
    52 -> 53 [ label = "CanProvide" ]
    53 -> 5 [ label = "Provides" ]
    52 -> 22 [ label = "CreateChildResolver" ]
    52 -> 25 [ label = "CreateChildResolver" ]
    51 -> 31 [ label = "CreateChildResolver" ]
    49 -> 54 [ label = "CanProvide" ]
    54 -> 3 [ label = "Provides" ]
    49 -> 19 [ label = "CreateChildResolver" ]
    54 -> 22 [ label = "CreateChildResolver" ]
    54 -> 27 [ label = "CreateChildResolver" ]
    54 -> 55 [ label = "CanProvide" ]
    55 -> 6 [ label = "Provides" ]
    49 -> 56 [ label = "CanProvide" ]
    56 -> 4 [ label = "Provides" ]
    49 -> 31 [ label = "CreateChildResolver" ]
    56 -> 57 [ label = "CanProvide" ]
    57 -> 7 [ label = "Provides" ]
    56 -> 35 [ label = "CreateChildResolver" ]
    56 -> 40 [ label = "CreateChildResolver" ]
    10 -> 16 [ label = "CreateChildResolver" ]
    10 -> 50 [ label = "CreateChildResolver" ]
    1 -> 58 [ label = "Field" ]
    50 -> 58 [ label = "Requires" ]
    10 -> 59 [ label = "CanProvide" ]
    59 -> 58 [ label = "Provides" ]
    12 -> 60 [ label = "CanProvide" ]
    60 -> 58 [ label = "Provides" ]
    14 -> 61 [ label = "CanProvide" ]
    61 -> 58 [ label = "Provides" ]
    2 -> 62 [ label = "Field" ]
    43 -> 62 [ label = "Requires" ]
    15 -> 63 [ label = "CanProvide" ]
    63 -> 62 [ label = "Provides" ]
    17 -> 64 [ label = "CanProvide" ]
    64 -> 62 [ label = "Provides" ]
    49 -> 65 [ label = "CanProvide" ]
    65 -> 62 [ label = "Provides" ]
    51 -> 66 [ label = "CanProvide" ]
    66 -> 62 [ label = "Provides" ]
    4 -> 67 [ label = "Field" ]
    40 -> 67 [ label = "Requires" ]
    30 -> 68 [ label = "CanProvide" ]
    68 -> 67 [ label = "Provides" ]
    32 -> 69 [ label = "CanProvide" ]
    69 -> 67 [ label = "Provides" ]
    47 -> 70 [ label = "CanProvide" ]
    70 -> 67 [ label = "Provides" ]
    56 -> 71 [ label = "CanProvide" ]
    71 -> 67 [ label = "Provides" ]
    35 -> 67 [ label = "Requires" ]
    33 -> 67 [ label = "Requires" ]
    31 -> 62 [ label = "Requires" ]
    3 -> 72 [ label = "Field" ]
    27 -> 72 [ label = "Requires" ]
    18 -> 73 [ label = "CanProvide" ]
    73 -> 72 [ label = "Provides" ]
    20 -> 74 [ label = "CanProvide" ]
    74 -> 72 [ label = "Provides" ]
    42 -> 75 [ label = "CanProvide" ]
    75 -> 72 [ label = "Provides" ]
    44 -> 76 [ label = "CanProvide" ]
    76 -> 72 [ label = "Provides" ]
    52 -> 77 [ label = "CanProvide" ]
    77 -> 72 [ label = "Provides" ]
    54 -> 78 [ label = "CanProvide" ]
    78 -> 72 [ label = "Provides" ]
    25 -> 72 [ label = "Requires" ]
    22 -> 72 [ label = "Requires" ]
    19 -> 62 [ label = "Requires" ]
    16 -> 58 [ label = "Requires" ]
}
